File: DSA_Problems/Algorithms/Dynamic_Programming/LeetCode/generateParentheses.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:

    void solve(int n, int left, int right, vector<string>& container, string res) {
        if (left == right && right == n) {
            container.push_back(res);
            return;
        }
        if (left < n)
            solve(n, left + 1, right, container, res + '(');
        if (right < left)
            solve(n, left, right + 1, container, res + ')');
    }

    vector<string> generateParenthesis(int n) {
        vector<string> res;
        solve(n, 0, 0, res, "");
        return res;
    }
};

int main() {
    // test case 1.
    int n = 3;
    Solution ob;
    vector<string> check;
    check = ob.generateParenthesis(n);
    for (auto i : check)
        cout << i << "\n";
	return 0;
}

File: LeetCode/String/Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = [set() for _ in range(n + 1)]
        res[0].add('')
        for i in range(1, n + 1):
            res[i].update(f'({j})' for j in res[i - 1])
            for j in range(1, i):
                res[i].update(f'{k}{r}' for k in res[j] for r in res[i - j])
        return sorted(list(res[n]))
